File: flash_usdt.py
#!/usr/bin/env python3
"""
Multi-Network USDT Token Interaction Script
Supports: ERC20 (Ethereum), TRC20 (TRON), BEP20 (Binance Smart Chain)

EDUCATIONAL PURPOSE ONLY
This script demonstrates how to interact with USDT tokens across different blockchain networks.
"""

import json
import sys
from typing import Dict, Optional
from decimal import Decimal

try:
    from web3 import Web3
    try:
        # Try newer import path (web3.py 6.0+)
        from web3.middleware import ExtraDataToPOAMiddleware as poa_middleware
    except ImportError:
        try:
            # Try older import path (web3.py < 6.0)
            from web3.middleware import geth_poa_middleware as poa_middleware
        except ImportError:
            # No PoA middleware available
            poa_middleware = None
    WEB3_AVAILABLE = True
except ImportError:
    Web3 = None
    poa_middleware = None
    WEB3_AVAILABLE = False

try:
    from tronpy import Tron
    from tronpy.keys import PrivateKey
    TRONPY_AVAILABLE = True
except ImportError:
    Tron = None
    TRONPY_AVAILABLE = False


class NetworkConfig:
    """Network configurations for different blockchains"""
    
    # Ethereum Mainnet (ERC20)
    ETHEREUM = {
        'name': 'Ethereum',
        'rpc': 'https://eth.llamarpc.com',
        'chain_id': 1,
        'usdt_contract': '0xdAC17F958D2ee523a2206206994597C13D831ec7',
        'explorer': 'https://etherscan.io',
        'type': 'ERC20'
    }
    
    # Binance Smart Chain (BEP20)
    BSC = {
        'name': 'Binance Smart Chain',
        'rpc': 'https://bsc-dataseed.binance.org/',
        'chain_id': 56,
        'usdt_contract': '0x55d398326f99059fF775485246999027B3197955',
        'explorer': 'https://bscscan.com',
        'type': 'BEP20'
    }
    
    # TRON Network (TRC20)
    TRON = {
        'name': 'TRON',
        'rpc': 'https://api.trongrid.io',
        'usdt_contract': 'TR7NHqjeKQxGTCi8q8ZY4pL8otSzgjLj6t',
        'explorer': 'https://tronscan.org',
        'type': 'TRC20'
    }
    
    @classmethod
    def get_all_networks(cls):
        """Get all available networks"""
        return {
            'ethereum': cls.ETHEREUM,
            'bsc': cls.BSC,
            'tron': cls.TRON
        }


class USDTFlashTool:
    """Multi-network USDT interaction tool"""
    
    # ERC20/BEP20 Standard ABI (simplified)
    ERC20_ABI = json.loads('[{"constant":true,"inputs":[],"name":"name","outputs":[{"name":"","type":"string"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[],"name":"totalSupply","outputs":[{"name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[],"name":"decimals","outputs":[{"name":"","type":"uint8"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[{"name":"_owner","type":"address"}],"name":"balanceOf","outputs":[{"name":"balance","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[],"name":"symbol","outputs":[{"name":"","type":"string"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"name":"_to","type":"address"},{"name":"_value","type":"uint256"}],"name":"transfer","outputs":[{"name":"","type":"bool"}],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":true,"inputs":[{"name":"_owner","type":"address"},{"name":"_spender","type":"address"}],"name":"allowance","outputs":[{"name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"}]')
    
    def __init__(self):
        """Initialize the USDT Flash Tool"""
        self.networks = NetworkConfig.get_all_networks()
        self.web3_instances = {}
        self.tron_instance = None
        
    def connect_to_network(self, network: str) -> bool:
        """
        Connect to a specific blockchain network
        
        Args:
            network: Network name ('ethereum', 'bsc', 'tron')
            
        Returns:
            bool: True if connection successful
        """
        if network not in self.networks:
            print(f"Error: Unknown network '{network}'")
            return False
            
        config = self.networks[network]
        
        try:
            if config['type'] in ['ERC20', 'BEP20']:
                # Check if web3 is available
                if not WEB3_AVAILABLE or Web3 is None:
                    print(f"✗ {config['type']} support not available (web3 not installed)")
                    return False
                
                # Connect to EVM-compatible networks
                w3 = Web3(Web3.HTTPProvider(config['rpc']))
                
                # Add PoA middleware for BSC if available
                if network == 'bsc' and poa_middleware is not None:
                    try:
                        w3.middleware_onion.inject(poa_middleware, layer=0)
                    except Exception as e:
                        print(f"Warning: Could not inject PoA middleware: {str(e)}")
                
                if w3.is_connected():
                    self.web3_instances[network] = w3
                    print(f"✓ Connected to {config['name']}")
                    return True
                else:
                    print(f"✗ Failed to connect to {config['name']}")
                    return False
                    
            elif config['type'] == 'TRC20':
                # Connect to TRON network
                if not TRONPY_AVAILABLE or Tron is None:
                    print("✗ TRC20 support not available (tronpy not installed)")
                    return False
                    
                self.tron_instance = Tron()
                print(f"✓ Connected to {config['name']}")
                return True
                
        except Exception as e:
            print(f"✗ Error connecting to {config['name']}: {str(e)}")
            return False
            
    def get_balance(self, network: str, address: str) -> Optional[Decimal]:
        """
        Get USDT balance for an address on a specific network
        
        Args:
            network: Network name
            address: Wallet address
            
        Returns:
            Decimal: Balance in USDT, or None if error
        """
        if network not in self.networks:
            print(f"Error: Unknown network '{network}'")
            return None
            
        config = self.networks[network]
        
        try:
            if config['type'] in ['ERC20', 'BEP20']:
                if network not in self.web3_instances:
                    print(f"Error: Not connected to {config['name']}")
                    return None
                
                w3 = self.web3_instances[network]
                
                # Validate address
                if not w3.is_address(address):
                    print(f"Error: Invalid address format for {config['name']}")
                    return None
                
                # Get contract
                contract = w3.eth.contract(
                    address=w3.to_checksum_address(config['usdt_contract']),
                    abi=self.ERC20_ABI
                )
                
                # Get balance
                balance_wei = contract.functions.balanceOf(
                    w3.to_checksum_address(address)
                ).call()
                
                # Get decimals (USDT uses 6 decimals on most networks)
                decimals = contract.functions.decimals().call()
                balance = Decimal(balance_wei) / Decimal(10 ** decimals)
                
                return balance
                
            elif config['type'] == 'TRC20':
                if self.tron_instance is None:
                    print("Error: Not connected to TRON")
                    return None
                
                # Get TRC20 contract
                contract = self.tron_instance.get_contract(config['usdt_contract'])
                
                # Get balance
                balance_raw = contract.functions.balanceOf(address)
                decimals = contract.functions.decimals()
                balance = Decimal(balance_raw) / Decimal(10 ** decimals)
                
                return balance
                
        except Exception as e:
            print(f"Error getting balance: {str(e)}")
            return None
            
    def get_token_info(self, network: str) -> Optional[Dict]:
        """
        Get USDT token information on a specific network
        
        Args:
            network: Network name
            
        Returns:
            Dict: Token info (name, symbol, decimals, total supply)
        """
        if network not in self.networks:
            print(f"Error: Unknown network '{network}'")
            return None
            
        config = self.networks[network]
        
        try:
            if config['type'] in ['ERC20', 'BEP20']:
                if network not in self.web3_instances:
                    print(f"Error: Not connected to {config['name']}")
                    return None
                
                w3 = self.web3_instances[network]
                contract = w3.eth.contract(
                    address=w3.to_checksum_address(config['usdt_contract']),
                    abi=self.ERC20_ABI
                )
                
                info = {
                    'name': contract.functions.name().call(),
                    'symbol': contract.functions.symbol().call(),
                    'decimals': contract.functions.decimals().call(),
                    'contract': config['usdt_contract'],
                    'network': config['name'],
                    'type': config['type']
                }
                
                return info
                
            elif config['type'] == 'TRC20':
                if self.tron_instance is None:
                    print("Error: Not connected to TRON")
                    return None
                
                contract = self.tron_instance.get_contract(config['usdt_contract'])
                
                info = {
                    'name': contract.functions.name(),
                    'symbol': contract.functions.symbol(),
                    'decimals': contract.functions.decimals(),
                    'contract': config['usdt_contract'],
                    'network': config['name'],
                    'type': config['type']
                }
                
                return info
                
        except Exception as e:
            print(f"Error getting token info: {str(e)}")
            return None
            
    def display_network_info(self):
        """Display information about all supported networks"""
        print("\n" + "="*70)
        print("SUPPORTED NETWORKS AND USDT CONTRACTS")
        print("="*70)
        
        for network_key, config in self.networks.items():
            print(f"\n{config['name']} ({config['type']}):")
            print(f"  Contract: {config['usdt_contract']}")
            print(f"  Explorer: {config['explorer']}/address/{config['usdt_contract']}")
            print(f"  RPC: {config['rpc']}")


def print_banner():
    """Print application banner"""
    banner = """
╔═══════════════════════════════════════════════════════════════════╗
║                   MULTI-NETWORK USDT TOOL                         ║
║                   Educational Purpose Only                         ║
╚═══════════════════════════════════════════════════════════════════╝

Supported Networks:
  • Ethereum (ERC20)
  • Binance Smart Chain (BEP20)
  • TRON (TRC20)

Supported Wallets:
  • MetaMask
  • Trust Wallet
  • Web3 Compatible Wallets
  • Any wallet supporting the respective network

⚠️  DISCLAIMER:
This tool is for educational purposes only. Always verify transactions
and never share your private keys. Use at your own risk.
"""
    print(banner)


def print_menu():
    """Print main menu"""
    print("\n" + "="*70)
    print("MAIN MENU")
    print("="*70)
    print("1. Check USDT Balance")
    print("2. View Token Information")
    print("3. Display Network Information")
    print("4. Connect to All Networks")
    print("5. Exit")
    print("="*70)


def main():
    """Main application entry point"""
    # Check if dependencies are available
    if not WEB3_AVAILABLE and not TRONPY_AVAILABLE:
        print("ERROR: No blockchain libraries installed!")
        print("\nPlease install the required dependencies:")
        print("  pip install web3 tronpy python-dotenv")
        print("\nOr install from requirements.txt:")
        print("  pip install -r requirements.txt")
        sys.exit(1)
    
    if not WEB3_AVAILABLE:
        print("WARNING: web3 not installed. ERC20/BEP20 support disabled.")
        print("To enable: pip install web3\n")
    
    if not TRONPY_AVAILABLE:
        print("WARNING: tronpy not installed. TRC20 support disabled.")
        print("To enable: pip install tronpy\n")
    
    print_banner()
    
    tool = USDTFlashTool()
    
    while True:
        print_menu()
        choice = input("\nEnter your choice (1-5): ").strip()
        
        if choice == '1':
            # Check balance
            print("\nAvailable networks: ethereum, bsc, tron")
            network = input("Enter network: ").strip().lower()
            address = input("Enter wallet address: ").strip()
            
            if network not in tool.networks:
                print(f"Error: Invalid network '{network}'")
                continue
            
            # Connect if not already connected
            if network not in tool.web3_instances and network != 'tron':
                print(f"\nConnecting to {network}...")
                tool.connect_to_network(network)
            elif network == 'tron' and tool.tron_instance is None:
                print(f"\nConnecting to {network}...")
                tool.connect_to_network(network)
            
            print(f"\nFetching balance...")
            balance = tool.get_balance(network, address)
            
            if balance is not None:
                print(f"\n{'='*50}")
                print(f"Network: {tool.networks[network]['name']}")
                print(f"Address: {address}")
                print(f"USDT Balance: {balance:.6f} USDT")
                print(f"{'='*50}")
                
        elif choice == '2':
            # View token info
            print("\nAvailable networks: ethereum, bsc, tron")
            network = input("Enter network: ").strip().lower()
            
            if network not in tool.networks:
                print(f"Error: Invalid network '{network}'")
                continue
            
            # Connect if not already connected
            if network not in tool.web3_instances and network != 'tron':
                print(f"\nConnecting to {network}...")
                tool.connect_to_network(network)
            elif network == 'tron' and tool.tron_instance is None:
                print(f"\nConnecting to {network}...")
                tool.connect_to_network(network)
            
            print(f"\nFetching token information...")
            info = tool.get_token_info(network)
            
            if info:
                print(f"\n{'='*50}")
                print(f"Token Information - {info['network']}")
                print(f"{'='*50}")
                print(f"Name: {info['name']}")
                print(f"Symbol: {info['symbol']}")
                print(f"Decimals: {info['decimals']}")
                print(f"Type: {info['type']}")
                print(f"Contract: {info['contract']}")
                print(f"{'='*50}")
                
        elif choice == '3':
            # Display network info
            tool.display_network_info()
            
        elif choice == '4':
            # Connect to all networks
            print("\nConnecting to all networks...\n")
            for network in tool.networks.keys():
                tool.connect_to_network(network)
            print("\nConnection attempts completed.")
            
        elif choice == '5':
            # Exit
            print("\nThank you for using Multi-Network USDT Tool!")
            print("Stay safe and verify all transactions! 🔒\n")
            break
            
        else:
            print("Invalid choice. Please enter a number between 1 and 5.")
        
        input("\nPress Enter to continue...")


if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("\n\nExiting... Goodbye! 👋\n")
        sys.exit(0)
    except Exception as e:
        print(f"\nUnexpected error: {str(e)}")
        sys.exit(1)

File: flash-usdt.js
/**
 * Flash USDT Script - Multi-Network Support
 * 
 * This script demonstrates a flash loan concept using USDT (Tether USD)
 * across multiple blockchain networks: TRC20 (Tron), ERC20 (Ethereum), and BEP20 (BSC).
 * 
 * Flash loans allow users to borrow assets without collateral as long as 
 * the borrowed amount is returned within the same transaction.
 * 
 * Note: This is a demonstration script. In production, you would need:
 * - Connection to the respective network via Web3 or Ethers.js (or TronWeb for Tron)
 * - Actual smart contracts for flash loan providers (Aave, dYdX, etc.)
 * - Proper transaction handling and gas management
 */

// Wallet configurations
const WALLETS = {
    METAMASK: {
        name: 'MetaMask',
        type: 'browser-extension',
        icon: '🦊',
        supportedNetworks: ['ERC20', 'BEP20'], // Supports Ethereum and BSC
        features: ['DApp Browser', 'Token Swaps', 'NFT Support', 'Hardware Wallet Integration'],
        website: 'https://metamask.io',
        mobileApp: true,
        desktopApp: true,
        description: 'Most popular Ethereum wallet with extensive DeFi support'
    },
    TRUST_WALLET: {
        name: 'Trust Wallet',
        type: 'mobile-wallet',
        icon: '🛡️',
        supportedNetworks: ['TRC20', 'ERC20', 'BEP20'], // Supports all three
        features: ['Multi-Chain Support', 'DApp Browser', 'Staking', 'NFT Gallery'],
        website: 'https://trustwallet.com',
        mobileApp: true,
        desktopApp: false,
        description: 'Official Binance wallet with comprehensive multi-chain support'
    },
    BINANCE: {
        name: 'Binance Wallet',
        type: 'exchange-wallet',
        icon: '🟡',
        supportedNetworks: ['TRC20', 'ERC20', 'BEP20'], // Supports all three
        features: ['Exchange Integration', 'Low Fees', 'Instant Trading', 'Savings Products'],
        website: 'https://www.binance.com',
        mobileApp: true,
        desktopApp: true,
        description: 'Integrated exchange wallet with trading capabilities'
    },
    WEB3: {
        name: 'Web3 Wallet',
        type: 'generic-web3',
        icon: '🌐',
        supportedNetworks: ['TRC20', 'ERC20', 'BEP20'], // Generic Web3 support
        features: ['Smart Contract Interaction', 'Multiple Providers', 'Custom Networks', 'Developer Friendly'],
        website: 'https://web3js.org',
        mobileApp: false,
        desktopApp: true,
        description: 'Generic Web3 interface for blockchain interactions'
    }
};

// Network configurations
const NETWORKS = {
    TRC20: {
        name: 'Tron (TRC20)',
        symbol: 'TRX',
        chainId: 'Mainnet',
        rpcUrl: 'https://api.trongrid.io',
        usdtContract: 'TR7NHqjeKQxGTCi8q8ZY4pL8otSzgjLj6t',
        explorerUrl: 'https://tronscan.org',
        gasFee: 0.05, // TRX energy/bandwidth cost
        flashLoanFee: 0.05, // 0.05% - Lower fees on Tron
        avgBlockTime: 3, // seconds
        icon: '🔷'
    },
    ERC20: {
        name: 'Ethereum (ERC20)',
        symbol: 'ETH',
        chainId: 1,
        rpcUrl: 'https://mainnet.infura.io/v3/YOUR-PROJECT-ID',
        usdtContract: '0xdac17f958d2ee523a2206206994597c13d831ec7',
        explorerUrl: 'https://etherscan.io',
        gasFee: 0.15, // ETH gas cost estimate
        flashLoanFee: 0.09, // 0.09% - Standard flash loan fee
        avgBlockTime: 12, // seconds
        icon: '⟠'
    },
    BEP20: {
        name: 'Binance Smart Chain (BEP20)',
        symbol: 'BNB',
        chainId: 56,
        rpcUrl: 'https://bsc-dataseed.binance.org',
        usdtContract: '0x55d398326f99059fF775485246999027B3197955',
        explorerUrl: 'https://bscscan.com',
        gasFee: 0.008, // BNB gas cost estimate
        flashLoanFee: 0.05, // 0.05% - Lower fees on BSC
        avgBlockTime: 3, // seconds
        icon: '🟡'
    }
};

class FlashUSDTLoan {
    constructor(config = {}) {
        this.network = config.network || 'ERC20'; // Default to Ethereum
        this.networkConfig = NETWORKS[this.network];
        
        if (!this.networkConfig) {
            throw new Error(`Invalid network: ${this.network}. Valid options: TRC20, ERC20, BEP20`);
        }
        
        this.wallet = config.wallet || 'METAMASK'; // Default wallet
        this.walletConfig = WALLETS[this.wallet];
        
        if (!this.walletConfig) {
            throw new Error(`Invalid wallet: ${this.wallet}. Valid options: METAMASK, TRUST_WALLET, BINANCE, WEB3`);
        }
        
        // Check wallet-network compatibility
        if (!this.walletConfig.supportedNetworks.includes(this.network)) {
            console.warn(`⚠️  Warning: ${this.walletConfig.name} may have limited support for ${this.networkConfig.name}`);
        }
        
        this.loanAmount = config.loanAmount || 10000; // Default 10,000 USDT
        this.fee = config.fee || this.networkConfig.flashLoanFee; // Use network-specific fee
        this.profitTarget = config.profitTarget || 100; // Minimum profit in USDT
        this.walletAddress = config.walletAddress || this.generateMockAddress();
    }
    
    /**
     * Generate a mock wallet address for demonstration
     */
    generateMockAddress() {
        if (this.network === 'TRC20') {
            // Tron address format: T + 33 base58 characters
            const chars = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';
            let addr = 'T';
            for (let i = 0; i < 33; i++) {
                addr += chars.charAt(Math.floor(Math.random() * chars.length));
            }
            return addr;
        } else {
            // Ethereum/BSC address format: 0x + 40 hex characters
            let addr = '0x';
            for (let i = 0; i < 40; i++) {
                addr += Math.floor(Math.random() * 16).toString(16);
            }
            return addr;
        }
    }
    
    /**
     * Simulate wallet connection
     */
    async connectWallet() {
        console.log(`\n🔗 Connecting to ${this.walletConfig.icon} ${this.walletConfig.name}...`);
        console.log(`   Wallet Type: ${this.walletConfig.type}`);
        console.log(`   Address: ${this.walletAddress}`);
        console.log(`   Network: ${this.networkConfig.icon} ${this.networkConfig.name}`);
        
        // Simulate connection delay
        await new Promise(resolve => setTimeout(resolve, 100));
        
        console.log('   ✅ Wallet Connected Successfully!');
        
        return {
            connected: true,
            wallet: this.wallet,
            address: this.walletAddress,
            network: this.network
        };
    }
    
    /**
     * Get wallet information
     */
    getWalletInfo() {
        return {
            wallet: this.wallet,
            name: this.walletConfig.name,
            type: this.walletConfig.type,
            supportedNetworks: this.walletConfig.supportedNetworks,
            features: this.walletConfig.features,
            compatible: this.walletConfig.supportedNetworks.includes(this.network)
        };
    }
    
    /**
     * Get network information
     */
    getNetworkInfo() {
        return {
            network: this.network,
            name: this.networkConfig.name,
            symbol: this.networkConfig.symbol,
            chainId: this.networkConfig.chainId,
            usdtContract: this.networkConfig.usdtContract,
            gasFee: this.networkConfig.gasFee,
            flashLoanFee: this.networkConfig.flashLoanFee,
            avgBlockTime: this.networkConfig.avgBlockTime
        };
    }

    /**
     * Calculate the fee for the flash loan
     */
    calculateFee() {
        return this.loanAmount * (this.fee / 100);
    }

    /**
     * Calculate total repayment amount
     */
    calculateRepayment() {
        return this.loanAmount + this.calculateFee();
    }

    /**
     * Simulate borrowing USDT via flash loan
     */
    async borrowFlashLoan() {
        console.log(`🚀 Initiating Flash Loan on ${this.networkConfig.icon} ${this.networkConfig.name}...`);
        console.log(`   Via ${this.walletConfig.icon} ${this.walletConfig.name}`);
        console.log(`📊 Loan Amount: ${this.loanAmount} USDT`);
        console.log(`💰 Fee: ${this.calculateFee()} USDT (${this.fee}%)`);
        console.log(`⛽ Network Gas Fee: ~${this.networkConfig.gasFee} ${this.networkConfig.symbol}`);
        console.log(`💳 Total Repayment Required: ${this.calculateRepayment()} USDT`);
        console.log(`📍 USDT Contract: ${this.networkConfig.usdtContract}`);
        console.log(`👛 Wallet Address: ${this.walletAddress}`);
        
        return {
            borrowed: this.loanAmount,
            fee: this.calculateFee(),
            totalRepayment: this.calculateRepayment(),
            network: this.network,
            wallet: this.wallet,
            gasFee: this.networkConfig.gasFee
        };
    }

    /**
     * Simulate arbitrage operation
     * This is where the actual profit-making strategy would occur
     */
    async executeArbitrage(borrowedAmount) {
        console.log('\n⚡ Executing Arbitrage Strategy...');
        
        // Simulate arbitrage profit (in real scenario, this would involve:
        // - DEX trading
        // - Price differences across exchanges
        // - Actual token swaps)
        const profit = Math.random() * 200 + 50; // Random profit between 50-250 USDT
        
        console.log(`📈 Arbitrage completed`);
        console.log(`💵 Profit Generated: ${profit.toFixed(2)} USDT`);
        
        return borrowedAmount + profit;
    }

    /**
     * Simulate repaying the flash loan
     */
    async repayFlashLoan(availableAmount, requiredAmount) {
        console.log('\n💸 Repaying Flash Loan...');
        
        if (availableAmount >= requiredAmount) {
            const netProfit = availableAmount - requiredAmount;
            console.log('✅ Flash Loan Repaid Successfully!');
            console.log(`🎉 Net Profit: ${netProfit.toFixed(2)} USDT`);
            return {
                success: true,
                netProfit: netProfit
            };
        } else {
            const shortage = requiredAmount - availableAmount;
            console.log('❌ Insufficient funds to repay flash loan!');
            console.log(`⚠️ Shortage: ${shortage.toFixed(2)} USDT`);
            console.log('🔄 Transaction will be reverted.');
            return {
                success: false,
                shortage: shortage
            };
        }
    }

    /**
     * Execute the complete flash loan cycle
     */
    async execute() {
        console.log('═══════════════════════════════════════');
        console.log(`    FLASH USDT LOAN EXECUTION`);
        console.log(`    Network: ${this.networkConfig.icon} ${this.networkConfig.name}`);
        console.log(`    Wallet: ${this.walletConfig.icon} ${this.walletConfig.name}`);
        console.log('═══════════════════════════════════════');

        try {
            // Step 0: Connect wallet
            await this.connectWallet();
            
            // Step 1: Borrow flash loan
            const loanDetails = await this.borrowFlashLoan();

            // Step 2: Execute arbitrage strategy
            const finalAmount = await this.executeArbitrage(loanDetails.borrowed);

            // Step 3: Repay flash loan
            const result = await this.repayFlashLoan(finalAmount, loanDetails.totalRepayment);

            console.log('\n═══════════════════════════════════════');
            if (result.success) {
                console.log(`     ✅ FLASH LOAN SUCCESSFUL`);
                console.log(`     Wallet: ${this.walletConfig.icon} ${this.wallet}`);
                console.log(`     Network: ${this.networkConfig.icon} ${this.network}`);
                console.log(`     Net Profit: ${result.netProfit.toFixed(2)} USDT`);
            } else {
                console.log('     ❌ FLASH LOAN FAILED');
                console.log('     Transaction Reverted');
            }
            console.log('═══════════════════════════════════════\n');

            return {
                ...result,
                network: this.network,
                networkName: this.networkConfig.name,
                wallet: this.wallet,
                walletName: this.walletConfig.name
            };

        } catch (error) {
            console.error('❌ Error during flash loan execution:', error.message);
            return {
                success: false,
                error: error.message,
                network: this.network,
                wallet: this.wallet
            };
        }
    }
}

/**
 * Display wallet information and features
 */
function displayWalletInfo(walletType) {
    const wallet = WALLETS[walletType];
    console.log(`\n${wallet.icon} ${wallet.name}`);
    console.log(`   Type: ${wallet.type}`);
    
    // Build network list with validation
    const networkList = wallet.supportedNetworks
        .filter(n => NETWORKS[n]) // Only include valid networks
        .map(n => NETWORKS[n].icon + ' ' + n)
        .join(', ');
    console.log(`   Networks: ${networkList}`);
    
    console.log(`   Features:`);
    wallet.features.forEach(feature => console.log(`      • ${feature}`));
    console.log(`   Website: ${wallet.website}`);
    console.log(`   Mobile: ${wallet.mobileApp ? '✅' : '❌'}  Desktop: ${wallet.desktopApp ? '✅' : '❌'}`);
}

/**
 * Compare all available wallets
 */
async function compareWallets() {
    console.log('\n╔═══════════════════════════════════════════════════════════╗');
    console.log('║              WALLET COMPARISON                             ║');
    console.log('╚═══════════════════════════════════════════════════════════╝');
    
    for (const walletType of ['METAMASK', 'TRUST_WALLET', 'BINANCE', 'WEB3']) {
        displayWalletInfo(walletType);
    }
    
    console.log('\n💡 Recommendations:');
    console.log('   • For Ethereum DeFi: MetaMask');
    console.log('   • For Multi-Chain: Trust Wallet or Binance Wallet');
    console.log('   • For Trading: Binance Wallet');
    console.log('   • For Developers: Web3 Wallet\n');
}

/**
 * Display wallet-network compatibility matrix
 */
function displayCompatibilityMatrix() {
    console.log('\n╔═══════════════════════════════════════════════════════════╗');
    console.log('║        WALLET-NETWORK COMPATIBILITY MATRIX                 ║');
    console.log('╚═══════════════════════════════════════════════════════════╝\n');
    
    // Header
    console.log('Wallet                 │ TRC20  │ ERC20  │ BEP20  │');
    console.log('───────────────────────┼────────┼────────┼────────┤');
    
    // Rows - using fixed width to accommodate emoji + name
    Object.keys(WALLETS).forEach(walletType => {
        const wallet = WALLETS[walletType];
        const trc20 = wallet.supportedNetworks.includes('TRC20') ? '  ✅  ' : '  ❌  ';
        const erc20 = wallet.supportedNetworks.includes('ERC20') ? '  ✅  ' : '  ❌  ';
        const bep20 = wallet.supportedNetworks.includes('BEP20') ? '  ✅  ' : '  ❌  ';
        
        // Use fixed width of 21 to handle emoji properly
        const walletName = (wallet.icon + ' ' + wallet.name).padEnd(21);
        console.log(`${walletName} │ ${trc20}│ ${erc20}│ ${bep20}│`);
    });
    
    console.log('───────────────────────┴────────┴────────┴────────┘\n');
}

/**
 * Compare flash loan opportunities across all networks
 */
async function compareNetworks(loanAmount = 50000) {
    const formattedAmount = loanAmount.toLocaleString();
    const amountPadding = Math.max(0, 20 - formattedAmount.length);
    const paddedAmount = formattedAmount + ' '.repeat(amountPadding);
    
    console.log('\n╔═══════════════════════════════════════════════════════════╗');
    console.log('║        FLASH LOAN NETWORK COMPARISON                      ║');
    console.log(`║        Loan Amount: ${paddedAmount} USDT${' '.repeat(11)}║`);
    console.log('╚═══════════════════════════════════════════════════════════╝\n');
    
    const results = [];
    
    for (const networkType of ['TRC20', 'ERC20', 'BEP20']) {
        const flashLoan = new FlashUSDTLoan({
            network: networkType,
            loanAmount: loanAmount
        });
        
        const networkInfo = flashLoan.getNetworkInfo();
        const fee = flashLoan.calculateFee();
        
        console.log(`${NETWORKS[networkType].icon} ${networkInfo.name}`);
        console.log(`   Flash Loan Fee: ${fee.toFixed(2)} USDT (${networkInfo.flashLoanFee}%)`);
        console.log(`   Gas Fee: ~${networkInfo.gasFee} ${networkInfo.symbol}`);
        console.log(`   Block Time: ~${networkInfo.avgBlockTime}s`);
        console.log(`   Contract: ${networkInfo.usdtContract}`);
        console.log('');
        
        results.push({
            network: networkType,
            name: networkInfo.name,
            fee: fee,
            gasFee: networkInfo.gasFee,
            symbol: networkInfo.symbol,
            totalCost: fee
        });
    }
    
    // Find the cheapest network
    results.sort((a, b) => a.totalCost - b.totalCost);
    console.log(`💡 Most Cost-Effective: ${NETWORKS[results[0].network].icon} ${results[0].name}`);
    console.log(`   (Total Fee: ${results[0].fee.toFixed(2)} USDT + ${results[0].gasFee} ${results[0].symbol} gas)\n`);
    
    return results;
}

// Example usage
async function main() {
    // Display wallet information and compatibility
    console.log('\n' + '═'.repeat(60));
    console.log('WALLET & NETWORK INFORMATION');
    console.log('═'.repeat(60));
    
    await compareWallets();
    displayCompatibilityMatrix();
    
    // Compare networks
    await compareNetworks(50000);
    
    console.log('\n' + '═'.repeat(60));
    console.log('FLASH LOAN EXAMPLES WITH DIFFERENT WALLETS');
    console.log('═'.repeat(60) + '\n');
    
    // Example 1: MetaMask + Ethereum
    console.log('\n--- Example 1: MetaMask on Ethereum (ERC20) ---\n');
    const metamaskEth = new FlashUSDTLoan({
        wallet: 'METAMASK',
        network: 'ERC20',
        loanAmount: 50000
    });
    await metamaskEth.execute();
    
    // Example 2: Trust Wallet + BSC
    console.log('\n--- Example 2: Trust Wallet on Binance Smart Chain (BEP20) ---\n');
    const trustBsc = new FlashUSDTLoan({
        wallet: 'TRUST_WALLET',
        network: 'BEP20',
        loanAmount: 50000
    });
    await trustBsc.execute();
    
    // Example 3: Binance Wallet + Tron
    console.log('\n--- Example 3: Binance Wallet on Tron (TRC20) ---\n');
    const binanceTron = new FlashUSDTLoan({
        wallet: 'BINANCE',
        network: 'TRC20',
        loanAmount: 50000
    });
    await binanceTron.execute();
    
    // Example 4: Web3 Wallet + BSC
    console.log('\n--- Example 4: Web3 Wallet on Binance Smart Chain (BEP20) ---\n');
    const web3Bsc = new FlashUSDTLoan({
        wallet: 'WEB3',
        network: 'BEP20',
        loanAmount: 50000
    });
    await web3Bsc.execute();
}

// Run the script if executed directly
if (require.main === module) {
    main().catch(console.error);
}

// Export for use as a module
module.exports = { FlashUSDTLoan, NETWORKS, WALLETS, compareNetworks, compareWallets, displayCompatibilityMatrix };
